OpenEdXMobile/src/main/java/org/edx/mobile/util/DataBindingHelperUtils.kt

```diff
@@ -114,47 +123,109 @@ class DataBindingHelperUtils {
+                }
+                CourseDateType.NOT_YET_RELEASED -> {
+                    badgeBackground = android.R.color.transparent
+                    textAppearance = R.style.not_yet_released_badge_text_appearance
+                    badgeStrokeColor = R.color.tag_light_silver_border
+                }
+                else -> {
+                    return
+                }
+            }
+            addDateBadge(textView, courseDateType.getTitle(), badgeBackground, textAppearance, badgeIcon, badgeStrokeColor)
+            if (isToday && type == CourseDateType.COMPLETED) {
+                setBadge(textView, type, false, dateBlockItems)
+            }
+        }
+
+        private fun hasSameDateTypes(dateBlockItems: ArrayList<CourseDateBlock>?): Boolean {
+            if (dateBlockItems != null && dateBlockItems.isEmpty().not() && dateBlockItems.size > 1) {
+                val dateType = dateBlockItems[0].dateBlockBadge
+                for (i in 1 until dateBlockItems.size) {
+                    if (dateBlockItems[i].dateBlockBadge != dateType &&
+                            dateBlockItems[i].dateBlockBadge != CourseDateType.BLANK) {
+                        return false
                     }
                 }
             }
-            textView.visibility = View.VISIBLE
+            return true
+        }
+
+        /**
+         * Method to add the date badge at the end in given title with styles attributes
+         */
+        private fun addDateBadge(textView: TextView, title: String, badgeBackground: Int,
+                                 textAppearance: Int,
+                                 badgeIcon: IconDrawable?,
+                                 badgeStrokeColor: Int) {
+            // add badge title so can identify the actual badge position
+            val titleWithBadge = SpannableStringBuilder((textView.text as Spannable))
+            // add space before badge title
+            titleWithBadge.append("  $title")
+            textView.setText(titleWithBadge, TextView.BufferType.SPANNABLE)
+
+            // setup the date badge
+            val string = textView.text as Spannable
+            val chipDrawable = ChipDrawable.createFromResource(textView.context,
+                    R.xml.dates_badge_chip)
+            chipDrawable.text = title
+            chipDrawable.setChipBackgroundColorResource(badgeBackground)
+            chipDrawable.setTextAppearanceResource(textAppearance)
+            if (badgeIcon != null) {
+                chipDrawable.chipIcon = badgeIcon
+            }
+            if (badgeStrokeColor != -1) {
+                chipDrawable.setChipStrokeColorResource(badgeStrokeColor)
+                chipDrawable.strokeWidth = ParsingUtil.dpToPx(textView.context, 1.0F)
+            }
+            // Reduce the chip height to match with the provided design.
+            chipDrawable.setBounds(0, (chipDrawable.intrinsicHeight * .1).toInt(),
```

Comment:
IMO, we can reduce the chipDrawable spacing n size by this
`chipDrawable.setBounds(0, (chipDrawable.intrinsicHeight * -.12).toInt(),
                    chipDrawable.intrinsicWidth, (chipDrawable.intrinsicHeight * .6).toInt())`

Reply:
replaced by the following code to make parity with mockup and web.
```
// Reduce the chip height and vertical margins to match the design.
            chipDrawable.setBounds(0, (chipDrawable.intrinsicHeight * -0.15).toInt(),
                    chipDrawable.intrinsicWidth, (chipDrawable.intrinsicHeight * 0.675).toInt())
```